src/mode_fast.js

```diff
@@ -0,0 +1,54 @@
+'use strict';
+/* @flow */
+/* globals Map: false */
+
+/**
+ * The [mode](http://bit.ly/W5K4Yt) is the number that appears in a list the highest number of times.
+ * There can be multiple modes in a list: in the event of a tie, this
+ * algorithm will return the most recently seen mode.
+ *
+ * modeFast uses a Map object to keep track of the mode, instead of the approach
+ * used with `mode`, a sorted array. As a result, it is faster
+ * than `mode` and supports any data type that can be compared with `==`.
+ * It also requires a
+ * [JavaScript environment with support for Map](https://kangax.github.io/compat-table/es6/#test-Map),
+ * and will throw an error if Map is not available.
+ *
+ * This is a [measure of central tendency](https://en.wikipedia.org/wiki/Central_tendency):
+ * a method of finding a typical or central value of a set of numbers.
+ *
+ * @param {Array<*>} x input
+ * @returns {?*} mode
+ * @throws {ReferenceError} if the JavaScript environment doesn't support Map
+ * @example
+ * modeFast(['rabbits', 'rabbits', 'squirrels']); // => 'rabbits'
+ */
+function modeFast/*::<T>*/(x /*: Array<T> */)/*: ?T */ {
+
+    // This index will reflect the incidence of different values, indexing
+    // them like
+    // { value: count }
+    var index = new Map();
+
+    // A running `mode` and the number of times it has been encountered.
+    var mode;
+    var modeCount = 0;
+
+    for (var i = 0; i < x.length; i++) {
```

Comment:
Perf note: was wondering if caching `x.length` would give a detectable speedup. Tried it, and it didn't.